packages/react/src/components/TextArea/TextArea.js

```diff
@@ -6,41 +6,73 @@
  */
 
 import PropTypes from 'prop-types';
-import React from 'react';
+import React, { useState, useRef } from 'react';
 import classNames from 'classnames';
 import { settings } from 'carbon-components';
-import { WarningFilled16 } from '@carbon/icons-react';
+import { WarningFilled16, EditOff16 } from '@carbon/icons-react';
+import Tooltip from '../Tooltip';
+import requiredIfValueExists from '../../prop-types/requiredIfValueExists';
+import { useControlledStateWithValue } from '../../internal/FeatureFlags';
 
 const { prefix } = settings;
 
-const TextArea = React.forwardRef(function TextArea(
+const DefaultCharCounter = ({ disabled, count, maxLength }) => {
+  const charCounterClasses = classNames(
+    `${prefix}--text-area--character-counter`,
     {
-      className,
-      id,
-      labelText,
-      hideLabel,
-      onChange,
-      onClick,
-      invalid,
-      invalidText,
-      helperText,
-      light,
-      ...other
-    },
-    ref
-)  {
+      [`${prefix}--text-area--character-counter--disabled`]: disabled,
+    }
+  );
+  return (
+    <span className={charCounterClasses}>
+      <span className={`${prefix}--text-area--character-counter--length`}>
+        {count}
+      </span>
+      /
+      <span className={`${prefix}--text-area--character-counter--maxlength`}>
+        {maxLength}
+      </span>
+    </span>
+  );
+};
+
+const TextArea = React.forwardRef(function TextArea(
+  {
+    className,
+    id,
+    labelText,
+    hideLabel,
+    onChange,
+    onClick,
+    invalid,
+    invalidText,
+    helperText,
+    light,
+    useCharCount,
+    maxLength,
+    defaultValue,
+    readOnly,
+    readOnlyIconLabel,
+    renderCharCounter: CharCounter = DefaultCharCounter,
+    ...other
+  },
+  ref
+) {
+  const [textareaVal, setInput] = useState(defaultValue);
+  const isControlled = useRef(other.value !== undefined).current;
   const textareaProps = {
     id,
     onChange: evt => {
       if (!other.disabled) {
-        onChange(evt);
+        onChange && onChange(evt, { value: evt.target.value });
       }
     },
     onClick: evt => {
       if (!other.disabled) {
         onClick(evt);
       }
     },
+    maxLength: maxLength || null,
```

Comment:
What's the reasoning behind the `null` case? Is it from passing through to `DefaultCharCounter` and us not wanting to render `undefined`?

Reply:
yeah it also deals with the case where maxLength is 0